### PR TITLE
Put together basic debrief UI and connect to backend

### PR DESCRIPTION
**Pull request by SylvanM**
## Overview

I put together the basic UI for weekly debrief, as well as implemented the primitive functionality for gathering the latest debrief info from the backend.

## Changes Made

Well, now there's a new state variable of the type `WeeklyDebriefDisplayInfo` which just contains the necessary information to display to the user for the weekly debrief.

### Backend connectivity

Upon the view being loaded, a query is made to populate the contents of a `WeeklyDebriefDisplayInfo` which is fed into the "state" variable, where the view is then reloaded to display the new contents.

### The actual UI

I pretty much just chained together some ForEach constructions that show all the possible articles that are stored in the display info struct that was passed in from the state variable. I also made it so that tapping on the article in the DebriefArticleView will open a BrowserView of that article.

## Test Coverage

I did manual testing, at first creating dummy debrief objects to test the display. After that, I just used the backend connectivity to load actual debriefs.

## Next Steps

Next, it'd be cool to have a skeleton view for the debrief, and for the "top X% of user" text to be more accurate.

## Screenshots

<details>
  <summary>First Page</summary>

[IMG_6838](https://user-images.githubusercontent.com/32347646/160898402-0fff9c8d-2886-44d2-9b3b-0bf2347b4843.PNG)

</details>

<details>
  <summary>Random Debrief Article View</summary>

[IMG_6837](https://user-images.githubusercontent.com/32347646/160898405-494164d2-6fbd-46e0-8b79-c0d7ddabf242.PNG)

</details>

<details>
  <summary>Conclusion Slide</summary>

[IMG_6836](https://user-images.githubusercontent.com/32347646/160898407-eddbaf74-c62c-4f36-b791-a36386fd6502.PNG)

</details>

### CODE REVIEW

File: Volume/Views/WeeklyDebriefView.swift [hunk 1]
```diff
@@ -6,112 +6,251 @@
...
 
-//struct WeeklyDebriefView_Previews: PreviewProvider {
-//    static var previews: some View {
-//        WeeklyDebriefView()
-//    }
-//}
+extension WeeklyDebriefView {
+    private enum WeeklyDebriefViewState<Result> {
+        case loading, results(Result)
+    }
```
Comment (dishanest) [inline Volume/Views/WeeklyDebriefView.swift:241]: Instead of reimplementing this for `WeeklyDebriefView`, consider modifying `MainView.TabState` by adding a `.weeklyDebriefPopup` case & renaming `TabState` to `ScreenState`. Remember to design APIs for the most general-purpose usage.

**Review by dishanest — CHANGES_REQUESTED** [review]
Great first PR @SylvanM! Left a bunch of comments, mostly related to style conventions. Also, here are additional comments about GitHub conventions: 

- Request your reviewers! Otherwise nobody gets a notification to review your work. Remember to request your pod lead (me), your pod co-iOS dev (@NgaiJustin), your subteam lead (@amyqxhuang), and your external PR reviewers (check the `#ios` channel). 
- Shorten the descriptions of your changes in the PR README. Currently it's very verbose, but it could be shortened to about half its length. Consider using bullet points to get bits of info across more quickly. 
- When attaching screenshots, display them in-line instead of hyperlinking them. If you know HTML, you could even restrict their width to 300px so they can be viewed on more device types. See [these docs](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#images) for Markdown image syntax. P.S. your screenshots are labeled in the wrong order. 
- Name your commits formally & concisely, in this format: `verb-object`. Never use "I" in a commit message. For example, "Made Home View throw up a dummy weekly debrief view when there is no debrief data, although, I haven't quite got that view up and working just yet" can be shortened to "Implement WeeklyDebriefView display using placeholder data."
- If your commit doesn't work, don't commit it just yet. Each commit should be an incremental step when you get a small part of your full feature to work. If you have to write "I haven't quite got that ... working just yet," then the commit isn't ready. 
- In the future, name your branch with this format: `firstname/feature-name`. In this case, your branch could be named `sylvan/test-debrief-ui`.

File: Volume/Views/BrowserView.swift [hunk 2]
```diff
@@ -32,6 +32,11 @@ struct BrowserView: View {
         }
     }
     
+    init(initType: BrowserViewInitType, navigationSource: NavigationSource) {
+        self.initType = initType
+        self.navigationSource = navigationSource
+    }
+
```
[the remark this answers is not in the document]
Comment (Npikielny) [reply Volume/Views/BrowserView.swift:39]: Yeah you only need this if you define other initializers or if you have private variables!

File: Volume/Views/StatisticView.swift [hunk 3]
```diff
@@ -36,6 +36,30 @@ struct StatisticView: View {
...
+                    .scaledToFit()
+                    .foregroundColor(Color.volume.orange)
+                    .frame(height: 24)
+                    .padding(.trailing, 24)
+                HStack(spacing: 8, content: {
+                    SkeletonView()
+                        .font(.begumMedium(size: 16))
+                    SkeletonView()
+                        .font(.begumMedium(size: 36))
+                        .foregroundColor(Color.volume.orange)
```
Comment (Npikielny) [inline Volume/Views/StatisticView.swift:53]:
```suggestion
                        .foregroundColor(.volume.orange)
```

File: Volume/Views/WeeklyDebriefView.swift [hunk 4]
```diff
@@ -6,112 +6,251 @@
...
+                    .frame(width: 64, height: 3)
+                    .padding(10)
+                TabView {
                     VStack {
+                        Image("volume-logo")
+                            .resizable()
+                            .frame(width: 245, height: 75)
+                            .padding(.top, 24)
+                        Text(getWeekString())
+                            .padding([.top, .bottom], 32)
```
Comment (Npikielny) [inline Volume/Views/WeeklyDebriefView.swift:62]:
```suggestion
                            .padding(.vertical, 32)
```

File: Volume/Views/WeeklyDebriefView.swift [hunk 5]
```diff
@@ -6,112 +6,251 @@
...
+                            .font(.begumMedium(size: 16))
+                        Divider()
+                            .frame(width: 100)
+                            .padding(.bottom, 48)
+                        VStack(spacing: 24, content: {
```
Comment (Npikielny) [inline Volume/Views/WeeklyDebriefView.swift:67]: +1 On what Han said earlier about implicit closures.

File: Volume/Views/WeeklyDebriefView.swift [hunk 6]
```diff
@@ -6,112 +6,251 @@
...
+                            }
+                            StatisticView(image: "volume", leftText: "read", number: debriefDisplayInfo.numReadArticles, rightText: "articles")
+                            StatisticView(image: "shout-out", leftText: "gave", number: debriefDisplayInfo.numShoutouts, rightText: "shout-outs")
+                            StatisticView(image: "bookmark", leftText: "bookmarked", number: debriefDisplayInfo.numBookmarkedArticles, rightText: "articles")
+                            HStack {
+                                Text("You were among the top ")
+                                    .font(.begumRegular(size: 16)) +
+                                Text("10% ")
+                                    .font(.begumMedium(size: 36))
+                                    .foregroundColor(Color.volume.orange) +
```
Comment (Npikielny) [inline Volume/Views/WeeklyDebriefView.swift:81]:
```suggestion
                                    .foregroundColor(.volume.orange) +
```

File: Volume/Views/WeeklyDebriefView.swift [hunk 7]
```diff
@@ -6,112 +6,251 @@
...
+                                Spacer()
+                            }
+                            HStack {
+                                Text("Keep up the volume! 📣")
+                                    .font(.begumRegular(size: 16))
+                                    .frame(alignment: .leading)
+                                Spacer()
+                            }
+                        })
+                            .padding([.leading, .trailing], 48)
```
Comment (Npikielny) [inline Volume/Views/WeeklyDebriefView.swift:93]:
```suggestion
                            .padding(.horizontal, 48)
```

File: Volume/Views/WeeklyDebriefView.swift [hunk 8]
```diff
@@ -6,112 +6,251 @@
...
+                                    .frame(width: 245, height: 75)
+                                Text("Stay updated with Cornell student publications, all in one place")
+                                    .font(.begumRegular(size: 16))
+                            }
+                            .padding(.top, 200)
+                            Button {
+                                isOpen = false
+                            } label: {
+                                Text("Continue to Volume")
+                                    .foregroundColor(Color.volume.orange)
```
Comment (Npikielny) [inline Volume/Views/WeeklyDebriefView.swift:126]:
```suggestion
                                    .foregroundColor(.volume.orange)
```

File: Volume/Views/WeeklyDebriefView.swift [hunk 9]
```diff
@@ -6,112 +6,251 @@
...
+                                    .font(.helveticaBold(size: 16))
+                            }
+                            .frame(width: 184, height: 45, alignment: .center)
+                            .overlay(RoundedRectangle(cornerRadius: 26)
+                            .stroke(Color.volume.orange, lineWidth: 2))
```
Comment (Npikielny) [inline Volume/Views/WeeklyDebriefView.swift:131]: Might be able to remove the `Color` here too.

File: Volume/Views/WeeklyDebriefView.swift [hunk 10]
```diff
@@ -6,112 +6,251 @@
...
-        .onDisappear() {
+        .onDisappear {
             UIPageControl.appearance().currentPageIndicatorTintColor = nil
             UIPageControl.appearance().pageIndicatorTintColor = nil
         }
     }
+    
+    
+    // MARK: - Copying from BrowserView
+
```
Comment (Npikielny) [inline Volume/Views/WeeklyDebriefView.swift:158]: Delete new lines at beginning
```suggestion
    
    // MARK: - Copying from BrowserView
    
```

File: Volume/Views/WeeklyDebriefView.swift [hunk 11]
```diff
@@ -6,112 +6,251 @@
...
+                }
+            }
+            receiveValue: { debriefData in
+                let debrief = WeeklyDebrief(from: debriefData.user.weeklyDebrief)
+                fetchInfoFor(debrief: debrief)
+            }
+    }
+    
+    private func fetchInfoFor(debrief: WeeklyDebrief) {
+
```
Comment (Npikielny) [inline Volume/Views/WeeklyDebriefView.swift:175]: Remove new line

File: Volume/Views/WeeklyDebriefView.swift [hunk 12]
```diff
@@ -6,112 +6,251 @@
...
+                    }
+                }
+                receiveValue: { articleFields in
+                    let article = Article(from: articleFields.article!.fragments.articleFields)
+                    
+                    if readNotRandom {
+                        debriefDisplayInfo.readArticles.append(article)
+                    } else {
+                        debriefDisplayInfo.randomArticles.append(article)
+                    }
```
Comment (Npikielny) [inline Volume/Views/WeeklyDebriefView.swift:204]: Stylistic change–I'd ask Han
```suggestion
                    (readNotRandom ? debriefDisplayInfo.readArticles : debriefDisplayInfo.randomArticles).append(article)
```

File: Volume/Views/WeeklyDebriefView.swift (hunk 10 above)
Comment (Npikielny) [reply Volume/Views/WeeklyDebriefView.swift:158]: The rest of AppDev projects wouldn't have one at the end either, but it looks like that's the style convention here.

File: Volume/Views/WeeklyDebriefView.swift [hunk 13]
```diff
@@ -6,112 +6,251 @@
...
+            formatter.setLocalizedDateFormatFromTemplate("M/dd")
+            
+            let firstString = formatter.string(from: firstDay)
+            let lastString = formatter.string(from: lastDay)
+            message.append(contentsOf: ", " + firstString + " - " + lastString)
+        }
+        
+        return message
+    }
+
```
Comment (Npikielny) [inline Volume/Views/WeeklyDebriefView.swift:235]:
```suggestion

```

**Review by Npikielny — CHANGES_REQUESTED** [review]
Great first PR! Make sure to request all of your reviewers!

File: Volume/Models/WeeklyDebrief.swift [hunk 14]
```diff
@@ -26,4 +27,26 @@ struct WeeklyDebrief: Codable {
         readArticleIDs = weeklyDebrief.readArticles.map(\.fragments.articleFields.id)
         randomArticleIDs = weeklyDebrief.randomArticles.map(\.fragments.articleFields.id)
     }
+    
+    init(creationDate: Date, expirationDate: Date, numShoutouts: Int, numReadArticles: Int, numBookmarkedArticles: Int, readArticleIDs: [ArticleID], randomArticleIDs: [ArticleID]) {
+        self.creationDate = creationDate
+        self.expirationDate = expirationDate
+        self.numShoutouts = numShoutouts
+        self.numReadArticles = numReadArticles
+        self.numBookmarkedArticles = numBookmarkedArticles
```
Comment (NgaiJustin) [inline Volume/Models/WeeklyDebrief.swift:36]: Alphabetize

File: Volume/Models/WeeklyDebrief.swift [hunk 15]
```diff
@@ -26,4 +27,26 @@ struct WeeklyDebrief: Codable {
...
+        self.randomArticleIDs = randomArticleIDs
+    }
+    
+    static func dummyDebrief() -> WeeklyDebrief {
+        WeeklyDebrief(
+            creationDate: Date(),
+            expirationDate: Date(),
+            numShoutouts: 11,
+            numReadArticles: 22,
+            numBookmarkedArticles: 33,
```
Comment (NgaiJustin) [inline Volume/Models/WeeklyDebrief.swift:47]: Also alphabetize

File: Volume/Views/StatisticView.swift [hunk 16]
```diff
@@ -36,6 +36,30 @@ struct StatisticView: View {
...
+                SkeletonView()
+                    .scaledToFit()
+                    .foregroundColor(Color.volume.orange)
+                    .frame(height: 24)
+                    .padding(.trailing, 24)
+                HStack(spacing: 8, content: {
+                    SkeletonView()
+                        .font(.begumMedium(size: 16))
+                    SkeletonView()
+                        .font(.begumMedium(size: 36))
```
Comment (NgaiJustin) [inline Volume/Views/StatisticView.swift:52]: We are no longer using this font (licensing issue). Pull from main branch to see all the current fonts.

File: Volume/Views/StatisticView.swift [hunk 17]
```diff
@@ -36,6 +36,30 @@ struct StatisticView: View {
     }
 }
 
+extension StatisticView {
+    struct Skeleton: View {
+        var body: some View {
+            HStack {
+                SkeletonView()
+                    .scaledToFit()
+                    .foregroundColor(Color.volume.orange)
```
Comment (NgaiJustin) [inline Volume/Views/StatisticView.swift:45]: .volume.orange

File: Volume/Views/WeeklyDebriefView.swift [hunk 18]
```diff
@@ -6,112 +6,251 @@
...
+            case .loading:
+                BigReadArticleRow.Skeleton() // This is until we have a SkeletonView for WeeklyDebrief
+            case .results(let debriefDisplayInfo):
+                Capsule()
+                    .fill(Color.secondary)
+                    .frame(width: 64, height: 3)
+                    .padding(10)
+                TabView {
                     VStack {
+                        Image("volume-logo")
```
Comment (NgaiJustin) [inline Volume/Views/WeeklyDebriefView.swift:57]: Use Volume Image extension instead

File: Volume/Views/WeeklyDebriefView.swift [hunk 19]
```diff
@@ -6,112 +6,251 @@
...
+                            .resizable()
+                            .frame(width: 245, height: 75)
+                            .padding(.top, 24)
+                        Text(getWeekString())
+                            .padding([.top, .bottom], 32)
+                            .font(.begumMedium(size: 16))
```
Comment (NgaiJustin) [inline Volume/Views/WeeklyDebriefView.swift:63]: Fix font here too.

File: Volume/Views/WeeklyDebriefView.swift [hunk 20]
```diff
@@ -6,112 +6,251 @@
...
+                                Text("10% ")
+                                    .font(.begumMedium(size: 36))
+                                    .foregroundColor(Color.volume.orange) +
+                                Text("active readers last week!")
+                                    .font(.begumRegular(size: 16))
+                                Spacer()
+                            }
+                            HStack {
+                                Text("Keep up the volume! 📣")
+                                    .font(.begumRegular(size: 16))
```
Comment (NgaiJustin) [inline Volume/Views/WeeklyDebriefView.swift:88]: I'll stop commenting about the font, just be sure to replace them all.

File: Volume/Views/WeeklyDebriefView.swift [hunk 21]
```diff
@@ -6,112 +6,251 @@
...
+                            StatisticView(image: "shout-out", leftText: "gave", number: debriefDisplayInfo.numShoutouts, rightText: "shout-outs")
+                            StatisticView(image: "bookmark", leftText: "bookmarked", number: debriefDisplayInfo.numBookmarkedArticles, rightText: "articles")
+                            HStack {
+                                Text("You were among the top ")
+                                    .font(.begumRegular(size: 16)) +
+                                Text("10% ")
+                                    .font(.begumMedium(size: 36))
+                                    .foregroundColor(Color.volume.orange) +
+                                Text("active readers last week!")
+                                    .font(.begumRegular(size: 16))
```
Comment (NgaiJustin) [inline Volume/Views/WeeklyDebriefView.swift:83]: Fix font here too.

File: Volume/Views/WeeklyDebriefView.swift (hunk 9 above)
Comment (SylvanM) [reply Volume/Views/WeeklyDebriefView.swift:131]: That results in getting an error `Type 'ShapeStyle' has no member 'volume'`

**Review by NgaiJustin — APPROVED** [review]
Looks good, just remember to update the font to the current ones and use the Image constants

**Review by emarc0314 — APPROVED** [review]
Aside from all the comments already, looks good!

**Comment by SylvanM** [comment]
I re-merged with main, the CORRECT way (ty Justin and Han) and (I think) I resolved the PR comments.

**Review by dishanest — APPROVED** [review]
Nice changes! My next PR will be making this UI match Figma designs :)

File: Volume/Views/DebriefArticleView.swift [hunk 22]
```diff
@@ -74,20 +86,28 @@ struct DebriefArticleView: View {
...
+                            .resizable()
+                            .grayBackground()
+                            .aspectRatio(contentMode: .fill)
+                            .frame(width: 275, height: 275)
+                            .clipped()
+                    } else {
+                        WebImage(url: article.publication.profileImageUrl)
+                            .resizable()
+                            .grayBackground()
+                            .frame(width: 275, height: 275)
```
Comment (amyqxhuang) [inline Volume/Views/DebriefArticleView.swift:102]: factor out the constants into a variable eg.`.frame(width: frameWidth, height: frameHeight)`

File: Volume/Views/DebriefArticleView.swift [hunk 23]
```diff
@@ -74,20 +86,28 @@ struct DebriefArticleView: View {
...
-                        .grayBackground()
-                        .frame(width: 275, height: 275)
-                        .padding(30)
+            NavigationLink(destination: BrowserView(initType: .readyForDisplay(article), navigationSource: .weeklyDebrief)) {
+                VStack(spacing: 16) {
+                    if let url = article.imageUrl {
+                        WebImage(url: url)
+                            .resizable()
+                            .grayBackground()
+                            .aspectRatio(contentMode: .fill)
```
Comment (amyqxhuang) [inline Volume/Views/DebriefArticleView.swift:95]: nit: `.scaledToFill()`

File: Volume/Views/WeeklyDebriefView.swift [hunk 24]
```diff
@@ -30,21 +54,21 @@ struct WeeklyDebriefView: View {
...
+                    VStack(spacing: 24) {
                         HStack {
                             Text("In the past week, you...")
                                 .font(.newYorkRegular(size: 16))
                             Spacer()
                         }
-                        StatisticView(image: "feed", leftText: "read", number: 10, rightText: "articles")
-                        StatisticView(image: "shout-out", leftText: "gave", number: 45, rightText: "shout-outs")
-                        StatisticView(image: "bookmark", leftText: "bookmarked", number: 10, rightText: "articles")
+                        StatisticView(image: "volume", leftText: "read", number: debriefDisplayInfo.numReadArticles, rightText: "articles")
```
Comment (amyqxhuang) [inline Volume/Views/WeeklyDebriefView.swift:69]: can we change `StatisticView` to take in an image since there is a static file for images?

Comment (amyqxhuang) [inline Volume/Views/WeeklyDebriefView.swift:69]: talk to han + justin about factoring out strings into a separate static file similar to what is done for images. 

this would be a separate PR

File: Volume/Views/WeeklyDebriefView.swift [hunk 25]
```diff
@@ -61,62 +85,149 @@ struct WeeklyDebriefView: View {
...
+    private func asyncPopulateDisplayInfo(recievedArticles: Int = 0, ids: [ArticleID], readNotRandom: Bool, displayInfo: WeeklyDebriefDisplayInfo, completion: @escaping (WeeklyDebriefDisplayInfo) -> ()) {
+        if recievedArticles == ids.count {
+            completion(displayInfo)
+            return
+        }
+        
+        cancellableArticleQuery = Network.shared.publisher(for: GetArticleByIdQuery(id: ids[recievedArticles]))
+            .sink { completion in
+                if case let .failure(error) = completion {
+                    print(error.localizedDescription)
```
Comment (amyqxhuang) [inline Volume/Views/WeeklyDebriefView.swift:164]: follow the standard error printing format that han established for volume

you can see at [this PR](https://github.com/cuappdev/volume-ios/pull/79)

File: Volume/Views/WeeklyDebriefView.swift [hunk 26]
```diff
@@ -61,62 +85,149 @@ struct WeeklyDebriefView: View {
...
+    /**
+     * Creates a date string for the past week
+     */
+    private func getWeekString() -> String {
+        var message = "Your weekly debrief"
+        
+        let firstDay = weeklyDebrief.creationDate.simpleString
+        let lastDay = weeklyDebrief.expirationDate.simpleString
+        
+        message.append(contentsOf: ", " + firstDay + " - " + lastDay)
```
Comment (amyqxhuang) [inline Volume/Views/WeeklyDebriefView.swift:209]: `", \(firstDay) - \(lastDay)"`?

**Review by amyqxhuang — CHANGES_REQUESTED** [review]
looks good! just some feedback :)

**Comment by emarc0314** [comment]
Looks good 🙌

File: Volume/Views/StatisticView.swift [hunk 27]
```diff
@@ -36,7 +36,32 @@ struct StatisticView: View {
...
+                        .foregroundColor(.volume.orange)
+                    SkeletonView()
+                        .font(.newYorkMedium(size: 16))
+                })
+                Spacer()
+            }
+        }
+    }
+}
+
```
Comment (emarc0314) [inline Volume/Views/StatisticView.swift:62]: extra space can be deleted

File: Volume/Views/WeeklyDebriefView.swift [hunk 28]
```diff
@@ -6,19 +6,43 @@
...
+    @Binding private var openedURL: Bool
+    @Binding private var onOpenArticleUrl: String?
+    
+    let initType: WeeklyDebriefViewInitType
+    
+    @State private var cancellableArticleQuery: AnyCancellable?
+    @State private var cancellableReadMutation: AnyCancellable?
+    @State private var cancellableShoutoutMutation: AnyCancellable?
+    @State private var state: MainView.TabState<WeeklyDebriefDisplayInfo> = .loading
+
```
Comment (emarc0314) [inline Volume/Views/WeeklyDebriefView.swift:33]: spacing

**Review by emarc0314 — APPROVED** [review]
Aside from some small spacing, looks good to me 🙌

File: Volume/Views/WeeklyDebriefView.swift (hunk 26 above)
Comment (SylvanM) [reply Volume/Views/WeeklyDebriefView.swift:209]: oh so true

File: Volume/Views/DebriefArticleView.swift (hunk 23 above)
Comment (SylvanM) [reply Volume/Views/DebriefArticleView.swift:95]: I'm assuming that nit means to add it

Comment (SylvanM) [reply Volume/Views/DebriefArticleView.swift:95]: nvm

File: Volume/Views/WeeklyDebriefView.swift (hunk 24 above)
Comment (SylvanM) [reply Volume/Views/WeeklyDebriefView.swift:69]: Yep!

**Review by Npikielny — APPROVED** [review]
